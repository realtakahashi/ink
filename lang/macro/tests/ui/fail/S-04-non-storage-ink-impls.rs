use ink_lang as ink;

#[ink::contract(version = "0.1.0")]
mod non_storage_ink_impls {
    // This test ensures that ink! impl blocks are always
    // implemented on the only storage struct definition.

    #[ink(storage)]
    struct StorageStruct {}

    // This ink! impl block is okay.
    impl StorageStruct {
        #[ink(constructor)]
        fn constructo1() -> Self {}

        #[ink(message)]
        fn message1(&self) {}
    }

    // Missing the #[ink(storage)] attribute on purpose.
    struct NonStorageStruct {}

    // This ink! impl block is invalid in that it implements
    // the messages and constructors for a non-existing ink!
    // storage struct. We expect a failure here.
    impl NonStorageStruct {
        #[ink(constructor)]
        fn constructor2() -> Self {}

        #[ink(message)]
        fn message2(&self) {}
    }
}

fn main() {}
