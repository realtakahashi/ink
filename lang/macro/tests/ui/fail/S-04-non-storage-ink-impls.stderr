error: ink! impl blocks need to be implemented for the #[ink(storage)] struct
  --> $DIR/S-04-non-storage-ink-impls.rs:26:10
   |
26 |     impl NonStorageStruct {
   |          ^^^^^^^^^^^^^^^^
